firestore/integration_test_internal/src/bundle_test.cc

```diff
@@ -0,0 +1,335 @@
+  // success.
+  ASSERT_EQ(progresses.size(), 4);
+  EXPECT_THAT(progresses[0], InProgressWithLoadedDocuments(0));
+  EXPECT_THAT(progresses[1], InProgressWithLoadedDocuments(1));
+  EXPECT_THAT(progresses[2], InProgressWithLoadedDocuments(2));
+  VerifySuccessProgress(progresses[3]);
+  EXPECT_EQ(progresses[3], final_progress);
+
+  VerifyQueryResults(db);
+}
+
+TEST_F(BundleTest, CanDeleteFirestoreFromProgressUpdate) {
+  Firestore* db = TestFirestore();
+  auto bundle = CreateTestBundle(db);
+
+  std::promise<void> db_deleted;
+  std::vector<LoadBundleTaskProgress> progresses;
+  Future<LoadBundleTaskProgress> result =
+      db->LoadBundle(bundle, [this, db, &progresses, &db_deleted](
+                                 const LoadBundleTaskProgress& progress) {
+        progresses.push_back(progress);
+        // Delete firestore before the final progress.
+        if (progresses.size() == 3) {
+          DeleteFirestore(db);
+          db_deleted.set_value();
+        }
+      });
+
+  // Wait for the notification that the instance is deleted before verifying.
+  db_deleted.get_future().wait();
+
+  // This future is not completed, and returns back a nullptr for result when it
+  // times out.
+  EXPECT_EQ(Await(result), nullptr);
+
+  // 3 progresses will be reported: initial, document 1, document 2.
+  // Final progress update is missing because Firestore is deleted before that.
+  ASSERT_EQ(progresses.size(), 3);
+  EXPECT_THAT(progresses[0], InProgressWithLoadedDocuments(0));
+  EXPECT_THAT(progresses[1], InProgressWithLoadedDocuments(1));
+  EXPECT_THAT(progresses[2], InProgressWithLoadedDocuments(2));
+}
+
+TEST_F(BundleTest, LoadBundlesForASecondTimeSkips) {
+  // TODO(wuandy): This test fails on Windows CI, but
+  // local run is fine. We need to figure out why and re-enable it.
+  SKIP_TEST_ON_WINDOWS;
+
+  Firestore* db = TestFirestore();
+  auto bundle = CreateTestBundle(db);
+  LoadBundleTaskProgress first_load = AwaitResult(db->LoadBundle(bundle));
+  VerifySuccessProgress(first_load);
+
+  std::vector<LoadBundleTaskProgress> progresses;
+  LoadBundleTaskProgress second_load = AwaitResult(db->LoadBundle(
+      bundle, [&progresses](const LoadBundleTaskProgress& progress) {
+        progresses.push_back(progress);
+      }));
+
+  // There will be 4 progress updates if it does not skip loading.
+  ASSERT_EQ(progresses.size(), 1);
+  VerifySuccessProgress(progresses[0]);
+  EXPECT_EQ(progresses[0], second_load);
+
+  VerifyQueryResults(db);
+}
+
+TEST_F(BundleTest, LoadInvalidBundlesShouldFail) {
+  // TODO(wuandy): This test fails on Windows CI, but
+  // local run is fine. We need to figure out why and re-enable it.
+  SKIP_TEST_ON_WINDOWS;
+
+  Firestore* db = TestFirestore();
+  std::vector<std::string> invalid_bundles{
+      "invalid bundle obviously", "\"(╯°□°）╯︵ ┻━┻\"",
+      "\xc3\x28"  // random bytes
+  };
+  for (const auto& bundle : invalid_bundles) {
+    std::vector<LoadBundleTaskProgress> progresses;
```

Comment:
Using a `ScopedTrace` RAII object would be especially useful here to help determine which iteration of the loop failed.